src/node/entity_id.h

```diff
@@ -0,0 +1,167 @@
+    static constexpr size_t LENGTH = 64; // hex-encoded SHA-256 hash
+
+    using Value =
+      std::string; // < hex-encoded hash of caller's DER-encoded certificate
+    Value id;
+
+    EntityId() = default;
+    EntityId(const Value& id_) : id(id_) {}
+
+    void operator=(const EntityId& other)
+    {
+      id = other.id;
+    }
+
+    void operator=(const Value& id_)
+    {
+      id = id_;
+    }
+
+    bool operator==(const EntityId& other) const
+    {
+      return id == other.id;
+    }
+
+    bool operator!=(const EntityId& other) const
+    {
+      return !(*this == other);
+    }
+
+    bool operator<(const EntityId& other) const
+    {
+      return id < other.id;
+    }
+
+    operator std::string() const
+    {
+      return id;
+    }
+
+    auto& value()
+    {
+      return id;
+    }
+
+    auto& value() const
+    {
+      return id;
+    }
+
+    auto data() const
+    {
+      return id.data();
+    }
+
+    size_t size() const
+    {
+      return id.size();
+    }
+
+    MSGPACK_DEFINE(id);
+  };
+
+  inline void to_json(nlohmann::json& j, const EntityId& entity_id)
+  {
+    j = entity_id.id;
+  }
+
+  inline void from_json(const nlohmann::json& j, EntityId& entity_id)
+  {
+    if (j.is_string())
+    {
+      entity_id = j.get<std::string>();
+    }
+    else
+    {
+      throw JsonParseError(
+        fmt::format("Entity id should be hex-encoded string: {}", j.dump()));
+    }
+  }
+
+  inline std::string schema_name(const EntityId&)
+  {
+    return "EntityId";
+  }
+
+  inline void fill_json_schema(nlohmann::json& schema, const EntityId&)
+  {
+    schema["type"] = "string";
+
+    // According to the spec, "format is an open value, so you can use any
+    // formats, even not those defined by the OpenAPI Specification"
+    // https://swagger.io/docs/specification/data-models/data-types/#format
+    schema["format"] = "hex";
+  }
+
+  template <typename T>
+  void add_schema_components(T&, nlohmann::json& j, const EntityId&)
+  {
+    j["type"] = "string";
+    j["pattern"] = fmt::format("^[a-f0-9]{{{}}}$", EntityId::LENGTH);
+  }
+
+  using CallerId = EntityId;
+  using MemberId = EntityId;
+  using UserId = EntityId;
+  using NodeId = EntityId;
+}
+
+namespace std
+{
+  static inline std::ostream& operator<<(
+    std::ostream& os, const ccf::EntityId& entity_id)
+  {
+    os << entity_id.id;
+    return os;
+  }
+
+  template <>
+  struct hash<ccf::EntityId>
+  {
+    size_t operator()(const ccf::EntityId& entity_id) const
+    {
+      return std::hash<std::string>{}(entity_id.id);
+    }
+  };
+}
+
+static constexpr size_t entity_id_truncation_max_char_count = 10;
+
+FMT_BEGIN_NAMESPACE
+template <>
+struct formatter<ccf::EntityId>
+{
+  template <typename ParseContext>
+  auto parse(ParseContext& ctx)
+  {
+    return ctx.begin();
+  }
+
+  template <typename FormatContext>
+  auto format(const ccf::EntityId& entity_id, FormatContext& ctx)
+    -> decltype(ctx.out())
+  {
```

Comment:
Why are we systematically truncating those in format?

Reply:
I've removed the custom formatting and added an explicit `trim()` member which is called from `raft.h` when the node ID needs to be printed (only on "sunny-day" messages).